compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/dnd/ExternalDrag.desktop.kt

```diff
@@ -0,0 +1,419 @@
+        }
+    )
+
+    init {
+        addDropTargetListener(dragTargetListener)
+    }
+
+    override fun setActive(isActive: Boolean) {
+        super.setActive(isActive)
+        if (!isActive) {
+            windowDragCoordinates = null
+        }
+    }
+
+    /**
+     * Subscribes on drag events for [window].
+     * If drag is going and component is under pointer [onDragStart] can be called synchronously.
+     *
+     * @param componentCoordinates coordinates of the component used to properly detect when drag entered/exited component
+     * @return handler id that can be used later to remove subscription using [stopDragHandling]
+     */
+    fun installComponentDragHandler(
+        componentCoordinates: LayoutCoordinates,
+        onDragStart: (Offset) -> Unit,
+        onDrag: (Offset) -> Unit,
+        onDragCancel: () -> Unit,
+        onDrop: (DropData) -> Unit
+    ): Int {
+        isActive = true
+
+        handlers[idsCounter] = ComponentDragHandler(componentCoordinates, onDragStart, onDrag, onDragCancel, onDrop)
+
+        if (isExternalDragInsideComponent(componentCoordinates, windowDragCoordinates)) {
+            onDragStart(calculateOffset(componentCoordinates, windowDragCoordinates!!))
+        }
+        return idsCounter++
+    }
+
+    /**
+     * Unsubscribes handler with [handleId].
+     * Calls [ComponentDragHandler.onDragCancel] if drag is going and handler's component is under pointer
+     *
+     * Disable drag handling for [window] if there are no more handlers.
+     *
+     * @param handleId id provided by [installComponentDragHandler] function
+     */
+    fun stopDragHandling(handleId: Int) {
+        val handler = handlers[handleId]
+        if (handler != null && isExternalDragInsideComponent(handler.componentCoordinates, windowDragCoordinates)) {
+            handler.onDragCancel()
+        }
+        handlers.remove(handleId)
+
+        if (handlers.isEmpty()) {
+            isActive = false
+        }
+    }
+
+
+    private class ComponentDragHandler(
+        val componentCoordinates: LayoutCoordinates,
+        val onDragStart: (Offset) -> Unit,
+        val onDrag: (Offset) -> Unit,
+        val onDragCancel: () -> Unit,
+        val onDrop: (DropData) -> Unit
+    )
+
+    companion object {
+        private fun isExternalDragInsideComponent(
+            componentCoordinates: LayoutCoordinates?,
+            windowDragCoordinates: Offset?
+        ): Boolean {
+            if (componentCoordinates == null || windowDragCoordinates == null) {
+                return false
+            }
+
+            return componentCoordinates.boundsInWindow().contains(windowDragCoordinates)
+        }
+
+        private fun calculateOffset(
+            componentCoordinates: LayoutCoordinates,
+            windowDragCoordinates: Offset
+        ): Offset {
+            return componentCoordinates.windowToLocal(windowDragCoordinates)
+        }
+    }
+}
+
+// @VisibleForTesting
```

Comment:
Not sure if we can add annotations library to this gradle module. I've seen them only for android target

Reply:
VisibleForTesting needs for public functions only to share between modules.
In your case, it's redundant.

Reply:
Thanks! I often use it for classes/functions that can be private, but I have to provide internal visibility for tests. But no one in the module should use it except tests
If you think it is redundant, I'll remove it

Reply:
Yeah, also this is a better description of VisibleForTesting in doc comment:

```
/**
 * Denotes that the class, method or field has its visibility relaxed, so that it is more widely
 * visible than otherwise necessary to make code testable.
 * <p>
 * You can optionally specify what the visibility <b>should</b> have been if not for
 * testing; this allows tools to catch unintended access from within production
 * code.
 * <p>
 * Example:
 * <pre><code>
 *  &#64;VisibleForTesting(otherwise = VisibleForTesting.PROTECTED)
 *  public String printDiagnostics() { ... }
 * </code></pre>
 *
 * If not specified, the intended visibility is assumed to be private.
 */
```

Reply:
It means that it is better to provide this annotation)

Reply:
Sorry, maybe I am wrong